Update protocol buffer dependency for MacOS Catalina

yousefhamza opened the pull request:
This PR fixes the issues mentioned in https://github.com/microsoft/PLCrashReporter/issues/17

Updating protobuf alone seems to have fixed all the issues there.

Commit:
Replace old protobuf-c with newer version built for 64 bit

Commit:
Build with the new binary sources

Modify header search paths to see the new protobuf directory and use the
new protobuf-c.c to build the framework

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/signed)](https://cla.opensource.microsoft.com/microsoft/PLCrashReporter?pullRequest=19) <br/>All CLA requirements met.

yousefhamza commented:
I will close this for now to fix an issue with protobuf

Commit:
Fix dynamic loader failure in protoc

annakocheshkova commented:
@clpolet

MatkovIvan commented on CrashReporter.xcodeproj/project.pbxproj:
```diff
@@ -1300,89 +1307,105 @@
 				"${DERIVED_FILES_DIR}/${CURRENT_ARCH}/$(INPUT_FILE_BASE).pb-c.h",
 			);
-			script = "cd \"${INPUT_FILE_DIR}\" && \"${SRCROOT}/Dependencies/protobuf-2.0.3/bin/protoc-c\" --c_out=\"${DERIVED_FILES_DIR}/${CURRENT_ARCH}\" \"${INPUT_FILE_NAME}\"";
+			script = "cd \"${INPUT_FILE_DIR}\" && \"${SRCROOT}/Dependencies/protobuf-c-1.3.2/bin/protoc-c\" --c_out=\"${DERIVED_FILES_DIR}/${CURRENT_ARCH}\" \"${INPUT_FILE_NAME}\"\n";
```
There are 5 more places like this in the project's scripts (in test targets)

yousefhamza replied:
Fixed ✅

MatkovIvan commented on CrashReporter.xcodeproj/project.pbxproj:
```diff
@@ -5672,7 +5744,7 @@
 				GCC_PREPROCESSOR_DEFINITIONS = PLCR_PRIVATE;
 				GCC_WARN_ABOUT_MISSING_NEWLINE = YES;
-				HEADER_SEARCH_PATHS = "\"$(SRCROOT)/Dependencies/protobuf-2.0.3/include\"";
+				HEADER_SEARCH_PATHS = "\"$(SRCROOT)/Dependencies/protobuf-c-1.3.2/include\"/**";
```
No need to recursive pattern here

yousefhamza replied:
Fixed ✅

MatkovIvan commented on CrashReporter.xcodeproj/project.pbxproj:
```diff
@@ -508,7 +508,6 @@
 		05E732000EFA1AE3005EDFB7 /* PLCrashReporter.m in Sources */ = {isa = PBXBuildFile; fileRef = 05F40ACA0EF7379F008050CF /* PLCrashReporter.m */; };
 		05E732010EFA1AE3005EDFB7 /* PLCrashReport.m in Sources */ = {isa = PBXBuildFile; fileRef = 05F411A50EF8DA31008050CF /* PLCrashReport.m */; };
 		05E732020EFA1AE3005EDFB7 /* crash_report.proto in Sources */ = {isa = PBXBuildFile; fileRef = 059670C70EEFAC3A008A0601 /* crash_report.proto */; };
-		05E732030EFA1AE3005EDFB7 /* protobuf-c.c in Sources */ = {isa = PBXBuildFile; fileRef = 05F40F830EF850FC008050CF /* protobuf-c.c */; };
```
`protobuf-c.c`  doesn't included to test targets so it leads link errors

yousefhamza replied:
Fixed ✅

MatkovIvan commented on CrashReporter.xcodeproj/project.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings:
```diff
@@ -0,0 +1,10 @@
+<plist version="1.0">
+<dict>
+	<key>BuildSystemType</key>
+	<string>Original</string>
```
Could you please explain why do we need legacy build system now? It seems that it works well with default one

yousefhamza replied:
I returned to the old one in the middle because I was getting an error from the new build system "multiple actions produce the same output" but it's fixed 👍

MatkovIvan commented on Dependencies/protobuf-c-1.3.2/LICENSE:
```diff
@@ -0,0 +1,32 @@
+Copyright (c) 2008-2016, Dave Benson and the protobuf-c authors.
+All rights reserved.
```
TODO: need to check license stuff

MatkovIvan commented on Dependencies/protobuf-c-1.3.2/lib/libprotobuf.dylib:
```diff
@@ -0,0 +1 @@
+libprotobuf.18.dylib
```
It would be great if we can build binaries instead of copying (git submodule with some script or so). It allows easier to update it in the future.
Also, it would help to make sure how and from what it is built.
It's probably out of scope of this PR, but would be nice improvement.

yousefhamza replied:
Yes.
How about something like this https://github.com/homebrew/homebrew-bundle?

MatkovIvan reviewed (changes requested):
Hi @yousefhamza,

First of all - big thanks for your contribution, it's much appreciated!
I have a few points here - some of it is just to bringing attention to some parts or problems. Also, during reviewing and testing this I faced with a few other things that most likely was here before your changes. Please treat my comments here as just discussions or so.

I fixed some of my points during review so, feel free to pick it up 😃

In addition to comments to certain lines here, I have to mention:
* There are `protobuf-2.0.3/bin/protoc-c` references in the scripts (fixed in https://github.com/MatkovIvan/PLCrashReporter/commit/5b80a9ded8904d2eef54ba1adf0ba0a06205ffda)
* `protobuf-c.c` doesn't included to test targets so it leads link errors (fixed in https://github.com/MatkovIvan/PLCrashReporter/commit/5b80a9ded8904d2eef54ba1adf0ba0a06205ffda)
* More `protobuf_c_system_allocator` usage (fixed in https://github.com/MatkovIvan/PLCrashReporter/commit/d08cfbe66b476f564196e7e77a5fd0fefe9a9a5d)
* Since the compiler of proto files has major update (version 3) it has some warnings now - like `No syntax specified for the proto file: crash_report.proto. Please use 'syntax = "proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted to proto2 syntax.)`. I really hope that it's just a warnings because [v2 and v3 aren't compatible](https://www.ben-morris.com/handling-protocol-buffers-backwards-compatibility-between-versions-2-and-3-using-c/)
  - TODO: check if it can deserialized already saved crash reports
* Tests are completely broken now 😭, and it doesn't look related to this PR so it just the things that I faced:
  - Assert type mismatch at compile time (just commented out locally for now)
  - No member named '__pc' in 'struct __darwin_arm_thread_state64' (fixed in https://github.com/MatkovIvan/PLCrashReporter/commit/f74744ae17fc26691efd6449438e911ec2305040). So, it was missed in https://github.com/microsoft/PLCrashReporter/commit/9c516ddd900091b56019bc8f682e859f490b6376 and probably some of these macroses should be reused
  - macOS unit tests run via `RunMacOSUnitTests.sh` script that uses `/Developer/Tools/RunUnitTests` that was deleted a **while** ago. Also, it looks like `Tools/google-toolbox-for-mac-trunk-r582` is used **only** for this script. Updating it to [the latest](https://github.com/google/google-toolbox-for-mac) doesn't help. Probably we should consider to just remove it. Regarding the tests it looks like this script phase can be just removed (fixed in https://github.com/MatkovIvan/PLCrashReporter/commit/42fb932d6ae8ca11f35d3cff98fadec9c6002146)
  - A lot of triggered asserts and crashes in tests

iamclement commented on Dependencies/protobuf-2.0.3/README.txt:
```diff
@@ -1,12 +0,0 @@
-The binaries were build from unmodified protobuf 2.0.3 and protobuf-c 0.6 sources, targeted at:
-    - Mac OS X 10.5
-
-The source (see src/) directory, is an extraction of the protobuf-c runtime library. It
```
The protobuf sources were previously modified with some fixes, we should double check if these modifications are still needed in the new version.

yousefhamza replied:
Do we have to contact "Plausible Labs" for this? or protocol buffer maintainers?

iamclement replied:
It could also has been fixed already in the new version. But not sure where to look at.

Commit:
Fix test targets

Modified protoc ref in build rules in test targets, fixed usage of
protobuf_c_system_allocator, also fixed warning of protoc syntax.

yousefhamza commented:
Hi @MatkovIvan,

Thanks for taking the time to review the PR.

> There are protobuf-2.0.3/bin/protoc-c references in the scripts (fixed in MatkovIvan@5b80a9d)

Fixed ✅

> protobuf-c.c doesn't included to test targets so it leads link errors (fixed in MatkovIvan@5b80a9d)

Fixed ✅

> More protobuf_c_system_allocator usage (fixed in MatkovIvan@d08cfbe)

Fixed ✅

> Since the compiler of proto files has major update (version 3) it has some warnings now - like No syntax specified for the proto file: crash_report.proto. Please use 'syntax = "proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted to proto2 syntax.). I really hope that it's just a warnings because v2 and v3 aren't compatible

Fixed ✅

> TODO: check if it can deserialized already saved crash reports
How to test this?


> Tests are completely broken now 😭, and it doesn't look related to this PR so it just the things that I faced

Yea I had the same issues too

iamclement commented:
BTW @yousefhamza maybe a dumb question but why is this PR integrating Protobuf v3.7.1 while the latest version is v3.9.1?

yousefhamza commented:
@clpolet I was using brew to install protocol buffer and the latest version on it was 3.7.1

yousefhamza commented:
@clpolet @MatkovIvan Do you have more concerns about this?

iamclement commented:
@MatkovIvan is in vacation at the moment.
What's remaining is understanding what happens to existing crash serialized with old Protobuf version. If we can use the latest version easily that would be ideal. We'll need to update the licenses accordingly.

iamclement commented:
Looks like we can now install the very latest while we are at it. From homebrew:
```
protobuf: stable 3.9.1 (bottled), HEAD
Protocol buffers (Google's data interchange format)
https://github.com/protocolbuffers/protobuf/
Not installed
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/protobuf.rb
==> Dependencies
Build: autoconf ✔, automake ✘, libtool ✘, python ✘
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 41,972 (30 days), 91,185 (90 days), 381,088 (365 days)
install_on_request: 18,068 (30 days), 42,843 (90 days), 178,464 (365 days)
build_error: 0 (30 days)
```

yousefhamza commented:
@clpolet Sure 👍

lumaxis commented:
As you might [have seen](https://devblogs.microsoft.com/appcenter/plcrashreporter-maintainership-moving-to-microsoft/), we are going to be moving the official PLCrashReporter repo to the Microsoft org on GitHub.
That means, it will replace this fork. In the next minutes, I'm going to rename this repo and also archive it.

Since we are of course still interested in your contribution, please reopen the PR against the official repo. Thanks!

lumaxis commented:
While we figure out the best way to port the changes from this fork to the official repo, I unarchived this repo and we'd be happy to get this PR merged soon 🙂

annakocheshkova commented:
Hi there, please check out my comment with the latest updates [here](https://github.com/microsoft/plcrashreporter/issues/19#issuecomment-543711845). 
Closing this PR due to this.
